File: 438-find-all-anagrams-in-a-string/438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        
        out = []
        
        ht = {}
        
        left = 0
        matches = 0
        
        for i in range(len(p)):
            if p[i] not in ht:
                ht[p[i]] = 1
            else:
                ht[p[i]]+=1
        
        for r in range(len(s)):
            curChar = s[r]
            
            if curChar in ht:
                ht[curChar]-=1
                if ht[curChar] == 0:
                    matches+=1
            
            if matches == len(ht):
                out.append(left)
            
            if r >= len(p) - 1:
                leftChar = s[left]
                if leftChar in ht:
                    if ht[leftChar] == 0:
                        matches-=1
                    ht[leftChar]+=1
                left+=1
        
        return out
            
            
        

File: 438-find-all-anagrams-in-a-string/438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        
        unordered_map<char,int> hash;
        
        vector<int> out;
        
        int matches = 0;
        int left = 0;
        
        for(const auto& it: p) {
            hash[it]++;
        }
        
        for(int r = 0; r < s.size(); r++) {
            char curChar = s[r];
            
            if(hash.count(curChar)) {
                hash[curChar]--;
                if(hash[curChar] == 0) {
                    matches++;
                }
            }
            
            if(matches == hash.size()) {
                out.push_back(left);
            }
            
            if(r >= p.size() - 1) {
                char leftChar = s[left];
                if(hash.count(leftChar)) {
                    if(hash[leftChar] == 0) {
                        matches--;
                    }
                    hash[leftChar]++;
                }
                left++;
            }
        }
        
        return out;
        
    }
};